### PR TITLE
[LLVM][ARM] Latency mutations for cortex m55,m7 and m85

### PR DESCRIPTION
**Pull request by nasherm**
This patch adds latency mutations as a scheduling related speedup for the above mentioned cores. When benchmarking this pass on selected benchmarks we see a performance improvement of 1% on most benchmarks with some improving by up to 6%.

Author: David Penry <david.penry@arm.com>
Co-authored-by: Nashe Mncube <nashe.mncube@arm.com

### CODE REVIEW

File: llvm/lib/Target/ARM/ARMLatencyMutations.cpp [hunk 1]
```diff
@@ -0,0 +1,984 @@
...
+  CortexM7Overrides(const ARMBaseInstrInfo *TII, AAResults *AA)
+      : ARMOverrideBypasses(TII, AA) {
+    if (!DI)
+      DI.reset(new CortexM7InstructionInformation(TII));
+  }
+
+  void modifyBypasses(SUnit &) override;
+
+private:
+  static std::unique_ptr<InstructionInformation> DI;
```
Comment (davemgreen) [inline llvm/lib/Target/ARM/ARMLatencyMutations.cpp:607]: I have never been suse if this static that is initialized once and not freed would be acceptable with sanatizers and the like. Is there an alternative that hopefully wouldn't involve spending too much compile time re-calculating the info?

Comment (nasherm) [reply llvm/lib/Target/ARM/ARMLatencyMutations.cpp]: With my latest commit I moved DI (now named II) to the enclosing anonymous namespace. However, I think this effectively makes the variable global which could be less than ideal. But it's not exposed by a header or anything so it still preserves it's "private" status I believe

**Review by davemgreen — APPROVED** [review]
I worry that the allocation but not freeing the data will trip up some sanatizers. We might want to rewrite this so that we avoid the table entirely, just checking for the opcodes/data we need on the instructions directly.

If we ned that we can rewrite it, lets see. Otherwise this LGTM.

**Comment by kazutakahirata** [comment]
I'm getting the following warning:

```
llvm/lib/Target/ARM/ARMSubtarget.cpp:255:11: error: enumeration values 'CortexM55' and 'CortexM85' not handled in switch [-Werror,-Wswitch]
  switch (ARMProcFamily) {
          ^~~~~~~~~~~~~
```

Would you mind taking a look?  Thanks!

**Comment by dpenry** [comment]
You do need to add new subtargets to that switch statement when adding them.  These two should join the big list of CortexM and CortexR processors which have a simple break for their case.

**Comment by nasherm** [comment]
@kazutakahirata 

I believe I've fixed this in PR #116086

**Pull request merged**